src/wasi_unstable/mod.rs

```diff
@@ -526,99 +513,234 @@ pub fn path_symlink(old_path: &[u8], fd: Fd, new_path: &[u8]) -> Errno {
-    let mut so_datalen = MaybeUninit::<usize>::uninit();
-    unsafe {
-        (
-            __wasi_sock_send(
-                sock,
-                si_data.as_ptr(),
-                si_data.len(),
-                si_flags,
-                so_datalen.as_mut_ptr(),
-            ),
-            so_datalen.assume_init(),
-        )
-    }
+#[inline]
+pub fn sock_send(sock: Fd, si_data: &[CIoVec], si_flags: SiFlags) -> Result<usize, Error> {
+    unsafe_wrap!{ __wasi_sock_send(sock, si_data.as_ptr(), si_data.len(), si_flags) }
 }
 
-pub fn sock_shutdown(sock: Fd, how: SdFlags) -> Errno {
-    unsafe { __wasi_sock_shutdown(sock, how) }
+#[inline]
+pub fn sock_shutdown(sock: Fd, how: SdFlags) -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_sock_shutdown(sock, how) }
 }
 
-pub fn sched_yield() -> Errno {
-    unsafe { __wasi_sched_yield() }
+#[inline]
+pub fn sched_yield() -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_sched_yield() }
 }
 
-pub fn fd_prestat_get(fd: Fd) -> (Errno, Prestat) {
-    let mut buf = MaybeUninit::<Prestat>::uninit();
-    unsafe {
-        (
-            __wasi_fd_prestat_get(fd, buf.as_mut_ptr()),
-            buf.assume_init(),
-        )
-    }
+#[inline]
+pub fn fd_prestat_get(fd: Fd) -> Result<Prestat, Error> {
+    unsafe_wrap!{ __wasi_fd_prestat_get(fd) }
 }
 
-pub fn fd_prestat_dir_name(fd: Fd, path: &mut [u8]) -> Errno {
-    unsafe { __wasi_fd_prestat_dir_name(fd, path.as_mut_ptr(), path.len()) }
+#[inline]
+pub fn fd_prestat_dir_name(fd: Fd, path: &mut [u8]) -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_fd_prestat_dir_name(fd, path.as_mut_ptr(), path.len()) }
 }
 
-// TODO: Safe interfaces to the args and environ functions
-/*
-pub fn args_get(argv: *mut *mut u8, argv_buf: *mut u8) -> Errno {}
-pub fn args_sizes_get(argc: *mut usize, argv_buf_size: *mut usize) -> Errno {}
-pub fn environ_get(environ: *mut *mut u8, environ_buf: *mut u8) -> Errno {}
-pub fn environ_sizes_get(environ_count: *mut usize, environ_buf_size: *mut usize) -> Errno {}
-*/
+#[cfg(feature = "alloc")]
+pub fn get_args(mut process_arg: impl FnMut(&[u8])) -> Result<(), Error> {
+    use alloc::vec;
+
+    let (mut argc, mut argv) = (0, 0);
+    let ret = unsafe { __wasi_args_sizes_get (&mut argc, &mut argv) };
+    if let Some(err) = NonZeroU16::new(ret) { return Err(err); }
+
+    // TODO: remove allocations after stabilization of unsized rvalues, see:
+    // https://github.com/rust-lang/rust/issues/48055
+    let mut argc = vec![core::ptr::null_mut::<u8>(); argc];
+    let mut argv = vec![0u8; argv];
+    let ret = unsafe { __wasi_args_get(argc.as_mut_ptr(), argv.as_mut_ptr()) };
+    if let Some(err) = NonZeroU16::new(ret) { return Err(err); }
+
+    for ptr in argc {
+        for n in 0.. {
+            unsafe {
+                // replace with memchr?
+                if *ptr.offset(n as isize) == 0 {
+                    let slice = core::slice::from_raw_parts(ptr, n);
```

Comment:
`n` is offset relative to the `ptr`, so it can be used also as length. Let's take `b"abc\x00"`, the condition will be true for `n = 3`, which is equal to the length of our desired string.

Reply:
Ah, I misread what this code was doing. My preference would be to use `CStr` to do the work of finding NUL bytes, instead of manual scans.

Also, I do want to step back and look at the big picture here. WASI's current arg/environ handling is fairly grubby and will hopefully evolve. It feels too early to be working towards an abstract goal such as API completeness right now, because there's a bigger abstract goal of evolving WASI to a place where all this code can be simpler. The practical side effect of the change here seems to be making that more work, because we'll have more code to keep in sync between WASI, the `wasi` repo, and `libc` here.

Reply:
`CStr` is defined only in `std`, so we can't use it here.

Even if arg/environ API will change, I think it will be quite easy to update this crate. I guess most of the work will be on the `libc` side.